packages/grid/x-data-grid/src/models/gridFilterOperator.ts

```diff
@@ -3,13 +3,53 @@ import { GridFilterItem } from './gridFilterItem';
 import { GridCellParams } from './params/gridCellParams';
 import type { GridColDef } from './colDef/gridColDef';
 import type { GridValidRowModel } from './gridRows';
+import type { GridApiCommunity } from './api/gridApiCommunity';
+
+export type GridApplyFilterV7<R extends GridValidRowModel = any, V = any, F = V> = (
+  value: V,
+  row: R,
+  column: GridColDef<R, V, F>,
+  apiRef: React.MutableRefObject<GridApiCommunity>,
+) => boolean;
+
+type ApplyFilterInterface<R extends GridValidRowModel = any, V = any, F = V> =
+  | {
+      /**
+       * The callback that generates a filtering function for a given filter item and column.
+       * This function can return `null` to skip filtering for this item and column.
+       * @param {GridFilterItem} filterItem The filter item with which we want to filter the column.
+       * @param {GridColDef} column The column from which we want to filter the rows.
+       * @returns {null | ((params: GridCellParams) => boolean)} The function to call to check if a row pass this filter item or not.
+       */
+      getApplyFilterFn: (
+        filterItem: GridFilterItem,
+        column: GridColDef<R, V, F>,
+      ) => null | ((params: GridCellParams<R, V, F>) => boolean);
+    }
+  | {
+      /**
+       * The callback that generates a filtering function for a given filter item and column.
+       * This function can return `null` to skip filtering for this item and column.
+       * @param {GridFilterItem} filterItem The filter item with which we want to filter the column.
+       * @param {GridColDef} column The column from which we want to filter the rows.
+       * @returns {null | ((value: V, row: R, column: GridColDef<R, V, F>) => boolean)} The function to call to check if a row pass this filter item or not.
+       */
+      getApplyFilterFnV7: (
+        filterItem: GridFilterItem,
+        column: GridColDef<R, V, F>,
+      ) => null | GridApplyFilterV7<R, V, F>;
+    };
```

Comment:
For the filtering API changes, I've added `V7` functions. These use a new filtering API, while still allowing for the old (current) API for userland code.

This new API is similar for filter functions and quick filter functions. It's:
`(value: any, row: GridRowModel: column: GridColDef, apiRef: API) => boolean`

The big advantage is that there is no object allocation, the JS engine can pass all the arguments on the stack. There parameters also provide an escape hatch, should the filter function absolutely need the cell params or something else. One example of this is the quick filter function in `gridStringOperators.ts`, which need the `formattedValue`.

Once we make a major release, we can easily remove the old API & use the new V7 one.